### PR TITLE
Push all path related data into FlowCommandBuilder

### PR DESCRIPTION
**Pull request by surabujin**
To ensure data consistency remove any data collection into
FlowCommandBuilder and provide all required data in advance.

Some input data types have been reworked to simplify FlowCommandBuilder
it's API.

### CODE REVIEW

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/FlowProcessingAction.java [hunk 1]
```diff
@@ -98,10 +121,83 @@ protected FlowPath getFlowPath(PathId pathId) {
...
+
+        FlowEndpoint ingressEndpoint = FlowSideAdapter.makeIngressAdapter(flow, path).getEndpoint();
+        SwitchProperties switchProperties = getSwitchProperties(ingressEndpoint.getSwitchId());
+        pathView.removeCustomerPortRule(
+                checkIfSharedCustomerPortForwardingFlowCanBeRemoved(flow.getFlowId(), ingressEndpoint));
+        pathView.removeCustomerPortLldpRule(
+                checkIfSharedLldpPortRuleCanBeRemoved(flow.getFlowId(), ingressEndpoint, switchProperties));
+        pathView.removeCustomerPortArpRule(
+                checkIfSharedArpPortRuleCanBeRemoved(flow.getFlowId(), ingressEndpoint, switchProperties));
+        return pathView.build();
```
Comment (niksv) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/FlowProcessingAction.java]: it works wrong. With such checks update and reroute operations can delete needed shared rules. 

**Test to check:**
1. enable multiTable, switchLldp, switchArp properties for all switches
2. create a flow.
3. updated flow bandwidth via V2 update.
4. validate src and dst switches.

**Expected result:**
successful switch validation

**Actual results:**
3 missing rules (LLDP, ARP, inputCustomer)

**Reason:**
In all checkIfShare* methods you are checking that no other flow uses shared rules. But your current flow still need these shared rules. So you can't remove them

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/update/actions/RemoveOldRulesAction.java [hunk 2]
```diff
@@ -72,18 +68,4 @@ protected void perform(State from, State to, Event event, FlowUpdateContext cont
...
-        RequestedFlow originalFlow = stateMachine.getOriginalFlow();
-        RequestedFlow targetFlow = stateMachine.getTargetFlow();
-
-        return SpeakerRequestBuildContext.builder()
-                .removeCustomerPortRule(removeForwardCustomerPortSharedCatchRule(originalFlow, targetFlow))
-                .removeOppositeCustomerPortRule(removeReverseCustomerPortSharedCatchRule(originalFlow, targetFlow))
-                .removeCustomerPortLldpRule(removeForwardSharedLldpRule(originalFlow, targetFlow))
-                .removeOppositeCustomerPortLldpRule(removeReverseSharedLldpRule(originalFlow, targetFlow))
-                .removeCustomerPortArpRule(removeForwardSharedArpRule(originalFlow, targetFlow))
-                .removeOppositeCustomerPortArpRule(removeReverseSharedArpRule(originalFlow, targetFlow))
```
Comment (niksv) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/update/actions/RemoveOldRulesAction.java]: removeForward/Reverse methods are unused now, but they weren't removed

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/FlowProcessingAction.java [hunk 3]
```diff
@@ -82,6 +93,18 @@ protected Flow getFlow(String flowId, FetchStrategy fetchStrategy) {
                         format("Flow %s not found", flowId)));
     }
 
+    protected FlowPathSpeakerView getFlowPath(FlowPathSpeakerView pathSnapshot) {
+        return pathSnapshot.toBuilder()
+                .path(getFlowPath(pathSnapshot.getPath().getPathId()))
+                .build();
+    }
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/FlowProcessingAction.java:100]: Renamed.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/FlowProcessingAction.java [hunk 4]
```diff
@@ -98,10 +121,83 @@ protected FlowPath getFlowPath(PathId pathId) {
         return new HashSet<>(flowRepository.findFlowsIdsByEndpointWithMultiTableSupport(switchId, port));
     }
 
+    protected FlowPathSpeakerView makeFlowPathNewView(Flow flow, FlowPath path, PathResources resources)
+            throws ResourceAllocationException {
+        FlowPathSpeakerView.FlowPathSpeakerViewBuilder pathSnapshot = FlowPathSpeakerView.builder(path)
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/FlowProcessingAction.java:126]: They are used in the "complete" QinQ branch (removed).

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/FlowProcessingAction.java [hunk 5]
```diff
@@ -82,6 +93,18 @@ protected Flow getFlow(String flowId, FetchStrategy fetchStrategy) {
                         format("Flow %s not found", flowId)));
     }
 
+    protected FlowPathSpeakerView getFlowPath(FlowPathSpeakerView pathSnapshot) {
+        return pathSnapshot.toBuilder()
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/FlowProcessingAction.java:97]: correct.

By the idea, this "path view" is immutable storage for all objects related to the flow path required for `FlowCommandBuilder`. I can't remember why I had to make a path replacement inside these objects... looks like on some workaround.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/PathSwappingAction.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        extends FlowProcessingAction<T, S, E, C> {
+    private final FlowResourcesManager resourcesManager;
+
+    public PathSwappingAction(PersistenceManager persistenceManager, FlowResourcesManager resourcesManager) {
+        super(persistenceManager);
+        this.resourcesManager = resourcesManager;
+    }
+
+    protected void saveOldPrimaryPaths(
+            FlowPathSwappingFsm<?, ?, ?, ?> stateMachine, Flow flow, FlowPath forwardPath, FlowPath reversePath,
```
Comment (sergii-iakovenko) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/PathSwappingAction.java:44]: Shouldn't class's generic type T extends FlowPathSwappingFsm?

**Comment by surabujin** [comment]
Replaced with #3438

**Pull request closed without merging**